tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp

```diff
@@ -2353,6 +2456,130 @@ void ProcessCtorFunctions(llvm::Module &M, StringRef globalName,
+  if (!CA)
+    return;
+
+  SmallVector<Function *, 2> Ctors;
+  for (User::op_iterator i = CA->op_begin(), e = CA->op_end(); i != e; ++i) {
+    if (isa<ConstantAggregateZero>(*i))
+      continue;
+    ConstantStruct *CS = cast<ConstantStruct>(*i);
+    if (isa<ConstantPointerNull>(CS->getOperand(1)))
+      continue;
+
+    // Must have a function or null ptr.
+    if (!isa<Function>(CS->getOperand(1)))
+      continue;
+    Function *Ctor = cast<Function>(CS->getOperand(1));
+    DXASSERT(Ctor->getReturnType()->isVoidTy() && Ctor->arg_size() == 0,
+             "function type must be void (void)");
+
+    for (inst_iterator I = inst_begin(Ctor), E = inst_end(Ctor); I != E; ++I) {
+      if (CallInst *CI = dyn_cast<CallInst>(&(*I))) {
+        Function *F = CI->getCalledFunction();
+        DXASSERT(F->getReturnType()->isVoidTy(), "ctor must be void(void)");
+        DXASSERT(F->arg_empty(), "ctor must be void(void)");
+        Ctors.emplace_back(F);
+      } else {
+        DXASSERT(isa<ReturnInst>(&(*I)),
+                 "else invalid Global constructor function");
+      }
+    }
+  }
+
+  for (Function *Ctor : Ctors) {
+    SmallVector<GlobalVariable *, 2> ConstStaticGV;
+    bool bHasSideEffect = false;
+    // Check all StoreInst in Ctor which is store a GV has only one store.
+    for (inst_iterator I = inst_begin(Ctor), E = inst_end(Ctor); I != E; ++I) {
+      Instruction *Inst = cast<Instruction>(&(*I));
+      StoreInst *ST = dyn_cast<StoreInst>(Inst);
+      if (!ST) {
+        if (CallInst *CI = dyn_cast<CallInst>(Inst)) {
+          if (HasSideEffect(CI, typeSys)) {
+            bHasSideEffect = true;
+            break;
+          }
+        } else if (Inst->mayHaveSideEffects()) {
+          bHasSideEffect = true;
+          break;
+        }
+        continue;
+      }
+      GlobalVariable *GV = dyn_cast<GlobalVariable>(ST->getPointerOperand());
+      if (!GV)
+        continue;
+      if (!IsSingleStoreStaticGlobal(GV, typeSys)) {
+        // Store on GV with multiple defines.
+        bHasSideEffect = true;
+        break;
+      }
+      ConstStaticGV.emplace_back(GV);
+    }
+    // Not support case when Ctor has side effect.
+    if (bHasSideEffect)
+      continue;
+    // Only take care Function init 1 const static GV.
+    if (ConstStaticGV.size() != 1)
+      continue;
+    GlobalVariable *GV = ConstStaticGV.back();
+    // Clone Ctor.
+    FunctionType *FT =
+        FunctionType::get(Ctor->getReturnType(), {GV->getType()}, false);
+    Function *NewCtor = Function::Create(
+        FT, GlobalValue::LinkageTypes::InternalLinkage, Ctor->getName(), &M);
+
+    SmallVector<ReturnInst *, 2> Returns;
+    ValueToValueMapTy vmap;
+    // Map params.
+    // Add param for each const static GV to the cloned ctor.
+    // Change the store ptr to the param.
+    vmap[GV] = NewCtor->arg_begin();
+
+    llvm::CloneFunctionInto(NewCtor, Ctor, vmap, /*ModuleLevelChagnes*/ false,
```

Comment:
I'm concerned that an init function could read from another GV we replaced in an earlier iteration.  This has two problems:
1. that reference to the global will not be replaced with the alloca and will not be initialized
2. the init functions will be called in reverse order

Reply:
I just realized the second init function will have reference to first global replaced with local alloca and call to cloned first init function.  So the result should still be correct, though with redundant init that should hopefully be cleaned up with CSE.